### PR TITLE
Create block filters before poll interval

### LINKED ISSUE #151: evmconnect is unable to retrieve transaction receipts when block polling interval is set too long
`evmconnect` creates a new block filter after the block polling interval on the Blockchain client, before periodically polling it for changes. If we submit transactions before the block filter is created, we noticed that transaction receipt are not being fetched, and the transaction gets into an endless loop stuck in the pending state. The transaction gets confirmed if a subsequent transaction is being sent after the block polling interval.

Consensus: Clique 

Logs
```
[2024-08-27T09:04:33.936Z]  INFO evmconnect: Prepared transaction method=set(uint256) dataLen=36 gas=1000000 httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.936Z] DEBUG evmconnect: Listed 0 items httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.936Z] DEBUG evmconnect: RPC[000000002] --> eth_getTransactionCount httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.936Z] DEBUG evmconnect: ==> POST http://127.0.0.1:5100 breq=QucGSnB0 httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: <== POST http://127.0.0.1:5100 [200] (1.35ms) breq=QucGSnB0 httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.938Z]  INFO evmconnect: RPC[000000002] <-- eth_getTransactionCount [200] OK (1.40ms) httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: Wrote tx_0/ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: Next nonce 15 for signer 0xbd59d91df554d84b917192a47930164b3f2e69da spent httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: Read tx_0/ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: Wrote tx_0/ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.938Z]  INFO evmconnect: Tracking transaction ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E at nonce 0xbd59d91df554d84b917192a47930164b3f2e69da / 15 httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.938Z]  INFO evmconnect: <-- POST / [202] (2.50ms) httpreq=hmI3ENbT req=26SvAjxX
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: Listed 1 items role=policyloop
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: Inflight set updated with 1 additional transactions, length is now 1 head-id:ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E head-seq=01919314-6892-656e-8e76-e33d452fe737 tail-id:ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E tail-seq=01919314-6892-656e-8e76-e33d452fe737 old-tail= role=policyloop
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: Sending transaction ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E at nonce 0xbd59d91df554d84b917192a47930164b3f2e69da / 15 (lastSubmit=<nil>) role=policyloop
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: Gas Price json to parse: 0 role=policyloop
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: maxPriorityFeePerGas=0x<nil> maxFeePerGas=0x<nil> gasPrice=0x0 role=policyloop
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: gasPrice=0x0 role=policyloop
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: RPC[000000003] --> eth_sendTransaction role=policyloop
[2024-08-27T09:04:33.938Z] DEBUG evmconnect: ==> POST http://127.0.0.1:5100 breq=FJ7yXgt3 role=policyloop
[2024-08-27T09:04:33.941Z] DEBUG evmconnect: <== POST http://127.0.0.1:5100 [200] (2.97ms) breq=FJ7yXgt3 role=policyloop
[2024-08-27T09:04:33.942Z]  INFO evmconnect: RPC[000000003] <-- eth_sendTransaction [200] OK (3.03ms) role=policyloop
[2024-08-27T09:04:33.942Z]  INFO evmconnect: Transaction ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E at nonce 0xbd59d91df554d84b917192a47930164b3f2e69da / 15 submitted. Hash: 0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43 role=policyloop
[2024-08-27T09:04:33.942Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=1,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:04:33.942Z]  INFO evmconnect: Added pending item TX:th=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43 role=confirmations_receipts
[2024-08-27T09:04:33.942Z] DEBUG evmconnect: Read tx_0/ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E role=policyloop
[2024-08-27T09:04:33.942Z] DEBUG evmconnect: State transition to sub-status  role=policyloop
[2024-08-27T09:04:33.942Z] DEBUG evmconnect: Wrote tx_0/ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E role=policyloop
[2024-08-27T09:04:33.942Z] DEBUG evmconnect: Read tx_0/ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E role=policyloop
[2024-08-27T09:04:33.942Z] DEBUG evmconnect: Wrote tx_0/ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E role=policyloop
[2024-08-27T09:04:33.942Z] DEBUG evmconnect: Read tx_0/ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E role=policyloop
[2024-08-27T09:04:33.942Z] DEBUG evmconnect: Wrote tx_0/ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E role=policyloop
[2024-08-27T09:04:51.846Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:05:11.849Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:05:21.849Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:05:31.842Z] DEBUG evmconnect: RPC[000000004] --> eth_newBlockFilter role=blocklistener
[2024-08-27T09:05:31.842Z] DEBUG evmconnect: ==> POST http://127.0.0.1:5100 breq=FS2EIL2K role=blocklistener
[2024-08-27T09:05:31.844Z] DEBUG evmconnect: <== POST http://127.0.0.1:5100 [200] (1.30ms) breq=FS2EIL2K role=blocklistener
[2024-08-27T09:05:31.844Z]  INFO evmconnect: RPC[000000004] <-- eth_newBlockFilter [200] OK (1.34ms) role=blocklistener
[2024-08-27T09:05:31.844Z] DEBUG evmconnect: RPC[000000005] --> eth_getFilterChanges role=blocklistener
[2024-08-27T09:05:31.844Z] DEBUG evmconnect: ==> POST http://127.0.0.1:5100 breq=PNDlDMwb role=blocklistener
[2024-08-27T09:05:31.844Z] DEBUG evmconnect: <== POST http://127.0.0.1:5100 [200] (0.62ms) breq=PNDlDMwb role=blocklistener
[2024-08-27T09:05:31.844Z]  INFO evmconnect: RPC[000000005] <-- eth_getFilterChanges [200] OK (0.65ms) role=blocklistener
[2024-08-27T09:05:41.849Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:05:51.850Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:06:11.841Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:06:21.849Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:06:31.845Z] DEBUG evmconnect: RPC[000000006] --> eth_getFilterChanges role=blocklistener
[2024-08-27T09:06:31.845Z] DEBUG evmconnect: ==> POST http://127.0.0.1:5100 breq=lUfTVlu1 role=blocklistener
[2024-08-27T09:06:31.846Z] DEBUG evmconnect: <== POST http://127.0.0.1:5100 [200] (1.14ms) breq=lUfTVlu1 role=blocklistener
[2024-08-27T09:06:31.846Z]  INFO evmconnect: RPC[000000006] <-- eth_getFilterChanges [200] OK (1.19ms) role=blocklistener
[2024-08-27T09:06:41.842Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:06:51.850Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:07:11.849Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:07:31.848Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:07:31.848Z] DEBUG evmconnect: RPC[000000007] --> eth_getFilterChanges role=blocklistener
[2024-08-27T09:07:31.848Z] DEBUG evmconnect: ==> POST http://127.0.0.1:5100 breq=5pfTEid6 role=blocklistener
[2024-08-27T09:07:31.850Z] DEBUG evmconnect: <== POST http://127.0.0.1:5100 [200] (1.41ms) breq=5pfTEid6 role=blocklistener
[2024-08-27T09:07:31.850Z]  INFO evmconnect: RPC[000000007] <-- eth_getFilterChanges [200] OK (1.45ms) role=blocklistener
[2024-08-27T09:07:51.850Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:08:11.849Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:08:31.848Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:08:31.850Z] DEBUG evmconnect: RPC[000000008] --> eth_getFilterChanges role=blocklistener
[2024-08-27T09:08:31.850Z] DEBUG evmconnect: ==> POST http://127.0.0.1:5100 breq=BJzq47Z2 role=blocklistener
[2024-08-27T09:08:31.852Z] DEBUG evmconnect: <== POST http://127.0.0.1:5100 [200] (1.47ms) breq=BJzq47Z2 role=blocklistener
[2024-08-27T09:08:31.852Z]  INFO evmconnect: RPC[000000008] <-- eth_getFilterChanges [200] OK (1.57ms) role=blocklistener
[2024-08-27T09:08:51.850Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:09:11.849Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:09:21.850Z] DEBUG evmconnect: Policy engine executed for tx ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E (update=0,status=Pending,hash=0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43) role=policyloop
[2024-08-27T09:09:31.852Z] DEBUG evmconnect: RPC[000000009] --> eth_getFilterChanges role=blocklistener
[2024-08-27T09:09:31.852Z] DEBUG evmconnect: ==> POST http://127.0.0.1:5100 breq=UpwBY8GT role=blocklistener
[2024-08-27T09:09:31.854Z] DEBUG evmconnect: <== POST http://127.0.0.1:5100 [200] (1.68ms) breq=UpwBY8GT role=blocklistener
[2024-08-27T09:09:31.854Z]  INFO evmconnect: RPC[000000009] <-- eth_getFilterChanges [200] OK (1.74ms) role=blocklistener
[2024-08-27T09:09:41.847Z]  INFO evmconnect: Transaction ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E at nonce 0xbd59d91df554d84b917192a47930164b3f2e69da / 15 has not been mined after 307.90s role=policyloop
[2024-08-27T09:09:41.847Z] DEBUG evmconnect: Sending transaction ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E at nonce 0xbd59d91df554d84b917192a47930164b3f2e69da / 15 (lastSubmit=2024-08-27T09:04:33.942063915Z) role=policyloop
[2024-08-27T09:09:41.847Z] DEBUG evmconnect: Gas Price json to parse: 0 role=policyloop
[2024-08-27T09:09:41.847Z] DEBUG evmconnect: maxPriorityFeePerGas=0x<nil> maxFeePerGas=0x<nil> gasPrice=0x0 role=policyloop
[2024-08-27T09:09:41.847Z] DEBUG evmconnect: gasPrice=0x0 role=policyloop
[2024-08-27T09:09:41.847Z] DEBUG evmconnect: RPC[000000010] --> eth_sendTransaction role=policyloop
[2024-08-27T09:09:41.847Z] DEBUG evmconnect: ==> POST http://127.0.0.1:5100 breq=970RLN9t role=policyloop
[2024-08-27T09:09:41.849Z] DEBUG evmconnect: <== POST http://127.0.0.1:5100 [200] (2.20ms) breq=970RLN9t role=policyloop
[2024-08-27T09:09:41.849Z] ERROR evmconnect: RPC[000000010] <-- [200]: nonce too low role=policyloop
[2024-08-27T09:09:41.849Z] DEBUG evmconnect: Transaction ns1:904F177C-C790-4B01-BDF4-F2B4E52E629E at nonce 0xbd59d91df554d84b917192a47930164b3f2e69da / 15 known with hash: 0x33133a4d48f429444eded568f4af2931081b803254688a07eaed7c6cff0bcb43 (nonce too low) role=policyloop
```

### PR DESCRIPTION
**Pull request by rodion-lim-partior**
Fixes: #151 

Raising this PR to allow for block filters to be created on init instead of after the block polling interval.

### CODE REVIEW

**Comment by EnriqueL8** [comment]
Thanks for the PR @rodion-lim-partior  - @Chengxuan could you take a look at this one?

**Review by Chengxuan — CHANGES_REQUESTED** [review]
@rodion-lim-partior Thanks for spotting https://github.com/hyperledger/firefly-evmconnect/issues/151 !

The solution proposed is likely to make the problem happen less frequently as long as no transactions are submitted before the new block filter is created on the node.

However, there is still a race condition. I think a better fix here is to ensure the `checkStartedLocked` function doesn't return until a block filter is established.

Please also add a test to catch future regressions.

**Comment by rodion-lim-partior** [comment]
@Chengxuan, noticed the following when I tried to fix the race condition:

1. When we add a consumer to the block listener, we acquire a lock
2. Call `checkStartedLocked`, which starts the block listener loop in a separate goroutine
3. Establish block height next, which requires acquiring a lock to write to block listener, this is blocked till `checkStartedLocked` has finished
4. We then create a block filter. If I block `checkStartedLocked` until the filter is created, that ends up in a deadlock, since `checkStartedLocked` is required to finish to release the lock before establishing block height can proceed

**Comment by Chengxuan** [comment]
@rodion-lim-partior It feels the problem is within the scope of `checkStartedLocked` function, so the solution shouldn't need to touch the global lock that is used to protect shared resources for concurrency. 

The key problem to solve is synchronization between the  `block listener loop` separate go routine and the main go routine that created that separate go routine.

Using a channel can also be a solution and feels more idiomatic in Golang.

**Comment by rodion-lim-partior** [comment]
@Chengxuan, can you help to take a look at the current commit if this was what you referred to?

File: internal/ethereum/blocklistener.go [hunk 1]
```diff
@@ -185,6 +194,8 @@ func (bl *blockListener) listenLoop() {
 				log.L(bl.ctx).Errorf("Failed to establish new block filter: %s", err.Message)
 				failCount++
 				continue
+			} else {
+				close(bl.blockFilterEstablished)
```
Comment (Chengxuan) [inline internal/ethereum/blocklistener.go:198]: need a check here to make sure closing the channel is only attempted once.

**Comment by Chengxuan** [comment]
@rodion-lim-partior Changes look great! I pointed out one possible issue in the code, you should be able to produce that error case in a test.

Once tests are added, I'm happy to take another look. 

Thank you for the work!

**Comment by EnriqueL8** [comment]
Thank @Chengxuan and @rodion-lim-partior for the collaboration on this one! Looking forward to seeing some tests

**Comment by rodionlim** [comment]
@Chengxuan, the commit I pushed causes a deadlock in my local environment due to the reasons in my previous comments. I will have to shift the global locks downwards to resolve the deadlocks. From local testing, shifting the locks down does not seem to break the existing implementation. Let me know if that works, I will implement the tests and refactor based on your comments after.

**Comment by Chengxuan** [comment]
Hi @rodion-lim-partior I think you have a good understanding of the code now. So please feel free to try out different options and write tests. Please make sure the logic must ensure only 1 listener loop go routine can be created after any change. Feel free to split global mutex to be resource specific if needed.

**Comment by rodionlim** [comment]
@Chengxuan, got it. Will do some refactoring of the code and come back with the tests for another round of review. Thanks for the pointers!

File: internal/ethereum/blocklistener.go (hunk 1 above)
Comment (rodion-lim-partior) [reply internal/ethereum/blocklistener.go]: done now

**Comment by rodion-lim-partior** [comment]
@Chengxuan @EnriqueL8, apologies for the delay. I refactored the code a fair bit and fixed some of the race conditions along the way. Added the tests as well for the edge case and the happy case scenario.

Some observations (correct me if I'm wrong):
1. There are two functions `addConsumer` and `getHighestBlock` which used to lock `checkStartedLocked` itself, leading to behavior which is difficult to reason about and unintentionally cause certain logic to occur in a particular sequence (described in a previous comment)
2. There is a small chance that a consumer is not added to the block listener (because the thread is slow etc.), causing mock `getFilterChanges` calls in the tests to skip the dispatch call, resulting in failed tests. This has now been fixed as well
3. Now that we wait for block filters to be established, many of the tests in event_actions.go and event_listener.go need to mock the RPC calls to create the block filters and listen on changes, this has also been corrected.

File: internal/ethereum/blocklistener.go [hunk 2]
```diff
@@ -470,23 +518,28 @@ func (bl *blockListener) dispatchToConsumers(consumers []*blockUpdateConsumer, u
...
 }
 
-func (bl *blockListener) checkStartedLocked() {
+func (bl *blockListener) checkStartedLocked(ctx context.Context) {
 	if bl.listenLoopDone == nil {
 		bl.listenLoopDone = make(chan struct{})
 		go bl.listenLoop()
 	}
+
+	bl.blockTillBlockFilterEstablished(ctx)
```
Comment (EnriqueL8) [inline internal/ethereum/blocklistener.go:527]: why would we do this for every consumer? If the `listenLoop` is already running before. I think we only need this once

File: internal/ethereum/blocklistener.go [hunk 3]
```diff
@@ -470,23 +518,28 @@ func (bl *blockListener) dispatchToConsumers(consumers []*blockUpdateConsumer, u
...
 }
 
 func (bl *blockListener) addConsumer(c *blockUpdateConsumer) {
 	bl.mux.Lock()
-	defer bl.mux.Unlock()
-	bl.checkStartedLocked()
 	bl.consumers[*c.id] = c
+	bl.mux.Unlock()
+	bl.checkStartedLocked(context.Background())
```
Comment (EnriqueL8) [inline internal/ethereum/blocklistener.go:534]: We would propagate the correct context here

File: internal/ethereum/blocklistener.go (hunk 2 above)
Comment (EnriqueL8) [reply internal/ethereum/blocklistener.go:527]: This would block adding new consumers

File: internal/ethereum/blocklistener.go [hunk 4]
```diff
@@ -92,6 +100,34 @@ func newBlockListener(ctx context.Context, c *ethConnector, conf config.Section,
...
+		bl.blockFilterEstablished.isEstablished = false
+		bl.blockFilterEstablished.signal = make(chan struct{})
+	}
+}
+
+func (bl *blockListener) blockTillBlockFilterEstablished(ctx context.Context) {
+	select {
+	case <-bl.blockFilterEstablished.signal:
+	case <-bl.ctx.Done():
+	case <-ctx.Done():
```
Comment (EnriqueL8) [inline internal/ethereum/blocklistener.go:127]: not sure why we need both here

File: internal/ethereum/blocklistener.go [hunk 5]
```diff
@@ -185,6 +231,8 @@ func (bl *blockListener) listenLoop() {
 				log.L(bl.ctx).Errorf("Failed to establish new block filter: %s", err.Message)
 				failCount++
 				continue
+			} else {
```
Comment (EnriqueL8) [inline internal/ethereum/blocklistener.go:234]: nit: There is a continue above so no else needed

File: internal/ethereum/blocklistener.go (hunk 3 above)
Comment (EnriqueL8) [reply internal/ethereum/blocklistener.go:534]: and we need to keep it in the lock because we check the state of the block listener in checkStartedLocked with `bl.listenLoopDone == nil ` There could be a race condition where two loops get started if not

**Review by EnriqueL8 — COMMENTED** [review]
Looking good - a few thoughts

File: internal/ethereum/blocklistener.go (hunk 2 above)
Comment (rodion-lim-partior) [reply internal/ethereum/blocklistener.go:527]: within bl.blocktillblockfilterestablished, it checks if the channel is closed, if the loop is already running, the channel will be closed, this won't block

File: internal/ethereum/blocklistener.go (hunk 4 above)
Comment (rodion-lim-partior) [reply internal/ethereum/blocklistener.go:127]: context.Done is for client side timeout by functions like `listener.getInitialBlock`, `bl.ctx.Done` takes care of cases like a sigint

File: internal/ethereum/blocklistener.go (hunk 5 above)
Comment (rodion-lim-partior) [reply internal/ethereum/blocklistener.go:234]: good catch, thanks

File: internal/ethereum/blocklistener.go (hunk 3 above)
Comment (rodion-lim-partior) [reply internal/ethereum/blocklistener.go:537]: @EnriqueL8, added a patch to lock till the listenLoopDone is set

File: internal/ethereum/blocklistener.go (hunk 2 above)
Comment (EnriqueL8) [reply internal/ethereum/blocklistener.go:530]: We do if we put it inside the if block

Comment (rodionlim) [reply internal/ethereum/blocklistener.go:530]: Makes sense, I'll test it and shift it inside the if block if nothing breaks

Comment (Chengxuan) [reply internal/ethereum/blocklistener.go:530]: what @rodion-lim-partior has done here makes sense to me. Using a mutex to ensure that there will be only 1 instance of `listenLoop` to close the `signal` channel while ensuring all other go routines wait until the `signal` channel is closed matches the behavior we discussed.

File: internal/ethereum/blocklistener.go [hunk 6]
```diff
@@ -92,6 +100,34 @@ func newBlockListener(ctx context.Context, c *ethConnector, conf config.Section,
...
+func (bl *blockListener) unsetBlockFilterStatus() {
+	bl.blockFilterEstablished.mux.Lock()
+	defer bl.blockFilterEstablished.mux.Unlock()
+	if bl.blockFilterEstablished.isEstablished {
+		bl.blockFilterEstablished.isEstablished = false
+		bl.blockFilterEstablished.signal = make(chan struct{})
+	}
+}
+
+func (bl *blockListener) blockTillBlockFilterEstablished(ctx context.Context) {
```
Comment (Chengxuan) [inline internal/ethereum/blocklistener.go:123]: I suggest removing `ctx context.Context` from this function as it's not used in the way you described. @rodion-lim-partior To enable it to work in the way you described, you'll need to surface this extra parameter to the top level. For now, in the code, you just used `context.Background()` which only gets canceled when the program exit, seems a bit redundant to `bl.ctx.Done()`.

File: internal/ethereum/blocklistener.go [hunk 7]
```diff
@@ -136,22 +172,27 @@ func (bl *blockListener) establishBlockHeightWithRetry() error {
...
+
 		return false, nil
 	})
 }
 
 func (bl *blockListener) listenLoop() {
 	defer close(bl.listenLoopDone)
+	defer func() {
+		bl.unsetBlockFilterStatus()
+	}()
```
Comment (Chengxuan) [inline internal/ethereum/blocklistener.go:195]: should not need this function unless you'd like to introduce the behavior of un-initialising block listener which needs extra design itself.

File: internal/ethereum/blocklistener.go [hunk 8]
```diff
@@ -40,6 +40,12 @@ type blockUpdateConsumer struct {
 	updates chan<- *ffcapi.BlockHashEvent
 }
 
+type blockFilterStatus struct {
+	isEstablished bool
+	signal        chan struct{}
+	mux           sync.Mutex
```
Comment (Chengxuan) [inline internal/ethereum/blocklistener.go:46]: don't think this mutex is actually required as the upstream logic already ensures there is a single thread modify the channel, so a flag + a channel should does the job.

Can we name them consistently with existing methods name. e.g. `isStarted` & `startDone`?

**Comment by Chengxuan** [comment]
@rodionlim logic looks good to me 👍 , some minor comments to clean up bits that are not necessary.

**Comment by rodion-lim-partior** [comment]
@Chengxuan, pushed a commit to clean up the code, left with one pending comment that needs a review

**Comment by Chengxuan** [comment]
@rodion-lim-partior thanks. For `one pending comment that needs a review` I assume that's referring to the removal of ctx parameter in `blockTillBlockFilterEstablished` that you are reviewing and addressing. 

If there are something you'd like me to review, please let me know.

**Comment by rodion-lim-partior** [comment]
@EnriqueL8, yep that's right. Rest of your review comments have been addressed. The pending one that needs a review, I need @Chengxuan to take a look. Removing the context will cause test failures.

**Comment by Chengxuan** [comment]
@rodion-lim-partior I looked at the code closer and understood the need to have the context, here is the proposed change: https://github.com/partior-3p/firefly-evmconnect/pull/1

File: internal/ethereum/blocklistener.go (hunk 2 above)
Comment (EnriqueL8) [reply internal/ethereum/blocklistener.go:514]: Okay, so when I wrote this comment it was in conjunction with the call to `checkStartedLocked` still being locked by `bl.mux.Lock()` but yes if we are now removing it we would need a lock at this point

**Comment by Chengxuan** [comment]
Hi @rodion-lim-partior I'm planning to do a release to pick up another bug fix tomorrow. I'd like to get this PR in as well, please take a look at https://github.com/partior-3p/firefly-evmconnect/pull/1 and merge it if you are happy with it. I've sorted out the merge conflicts in that PR as well

**Comment by rodion-lim-partior** [comment]
@Chengxuan, merged those changes in. Corrected some of the lint failures as well when you run `make test`. Seems to be working fine on my local environment

**Review by Chengxuan — APPROVED** [review]
Awesome. Thanks for your contribution @rodion-lim-partior 👍

**Comment by rodion-lim-partior** [comment]
Cheer guys.
